### PR TITLE
WINDUP-2026: produce csv report of all applications' results merged

### PR DESCRIPTION
**Pull request by m-brophy**
There's another JIRA for providing a link in the web-ui to the file produced.
To check the file produced independently of that you need to navigate to where the reports for the analysis are stored in the file system.

### CODE REVIEW

**Comment by m-brophy** [comment]
Just noticed this other JIRA referenced doesn't exist so I'll wait to mark this ready to review until the web-ui part is done.

**Comment by m-brophy** [comment]
Link on All Issues report now included
screenshot from 2018-12-17 10-27-43

File: reporting/impl/src/main/java/org/jboss/windup/reporting/rules/generation/CreateMigrationIssuesReportRuleProvider.java [hunk 1]
```diff
@@ -86,12 +91,13 @@ private ApplicationReportModel createSingleApplicationMigrationIssuesReport(Grap
...
         }
 
-        private ApplicationReportModel createGlobalMigrationIssuesReport(GraphContext context)
+        private ApplicationReportModel createGlobalMigrationIssuesReport(GraphContext context, boolean isExportingCSV)
         {
             ReportService reportService = new ReportService(context);
             ApplicationReportModel report = createMigrationIssuesReportBase(context);
             report.setReportName(ALL_MIGRATION_ISSUES_REPORT_NAME);
             report.setDisplayInGlobalApplicationIndex(true);
+            report.setIsExportingCSV(isExportingCSV);
```
Comment (mrizzi) [inline reporting/impl/src/main/java/org/jboss/windup/reporting/rules/generation/CreateMigrationIssuesReportRuleProvider.java]:
```suggestion
            report.setExportAllIssuesCSV(exportAllIssuesCSV);
```

File: reporting/impl/src/main/java/org/jboss/windup/reporting/rules/generation/CreateMigrationIssuesReportRuleProvider.java [hunk 2]
```diff
@@ -86,12 +91,13 @@ private ApplicationReportModel createSingleApplicationMigrationIssuesReport(Grap
             return report;
         }
 
-        private ApplicationReportModel createGlobalMigrationIssuesReport(GraphContext context)
+        private ApplicationReportModel createGlobalMigrationIssuesReport(GraphContext context, boolean isExportingCSV)
```
Comment (mrizzi) [inline reporting/impl/src/main/java/org/jboss/windup/reporting/rules/generation/CreateMigrationIssuesReportRuleProvider.java]:
```suggestion
        private ApplicationReportModel createGlobalMigrationIssuesReport(GraphContext context, boolean exportAllIssuesCSV)
```

File: reporting/impl/src/main/resources/reports/templates/migration-issues.ftl [hunk 3]
```diff
@@ -134,7 +134,11 @@
                 <div class="page-header page-header-no-border">
                     <h1>
                         <div class="main">${reportModel.reportName}
-                        <i class="glyphicon glyphicon-info-sign" data-toggle="tooltip" data-placement=right title="${reportModel.description}"></i></div>
+                        <i class="glyphicon glyphicon-info-sign" data-toggle="tooltip" data-placement=right title="${reportModel.description}"></i>
+                        <#if !(reportModel.projectModel??) && reportModel.isExportingCSV>
+                            <a class="csvReport" style="font-size: 14pt;" align="right" href="../MergedAnalysis.csv">(CSV Report)</a>
```
Comment (mrizzi) [inline reporting/impl/src/main/resources/reports/templates/migration-issues.ftl]:
```suggestion
                            <a class="csvReport" style="font-size: 14pt;" align="right" href="../AllIssues.csv">Download all issues CSV `[with the patternfly icon pficon-export]`</a>
```
And please try to place the link right aligned just like in the mock-up.

File: reporting/impl/src/main/resources/reports/templates/migration-issues.ftl [hunk 4]
```diff
@@ -134,7 +134,11 @@
                 <div class="page-header page-header-no-border">
                     <h1>
                         <div class="main">${reportModel.reportName}
-                        <i class="glyphicon glyphicon-info-sign" data-toggle="tooltip" data-placement=right title="${reportModel.description}"></i></div>
+                        <i class="glyphicon glyphicon-info-sign" data-toggle="tooltip" data-placement=right title="${reportModel.description}"></i>
+                        <#if !(reportModel.projectModel??) && reportModel.isExportingCSV>
```
Comment (mrizzi) [inline reporting/impl/src/main/resources/reports/templates/migration-issues.ftl]:
```suggestion
                        <#if !(reportModel.projectModel??) && reportModel.isExportAllIssuesCSV>
```

File: reporting/impl/src/main/java/org/jboss/windup/reporting/export/ExportCSVFileRuleProvider.java [hunk 5]
```diff
@@ -66,63 +70,108 @@ public void perform(GraphRewrite event, EvaluationContext context, WindupConfigu
...
             final Iterable<ClassificationModel> classifications = classificationService.findAll();
+            List reportableEvents = new ArrayList<>();
+            reportableEvents.addAll((List<InlineHintModel>)hints);
+            reportableEvents.addAll((List<ClassificationModel>)classifications);
 
             //try{} in case something bad happens, we need to close files
             try
             {
-                for (InlineHintModel hint : hints)
+                reportableEvents.stream().sorted((o1,o2) ->
```
Comment (mrizzi) [inline reporting/impl/src/main/java/org/jboss/windup/reporting/export/ExportCSVFileRuleProvider.java]: i changed this method a little bit in https://gist.github.com/mrizzi/b30dea66197ff3db5ae486a12ff8a526

File: reporting/impl/src/main/java/org/jboss/windup/reporting/export/ExportCSVFileRuleProvider.java [hunk 6]
```diff
@@ -66,63 +70,108 @@ public void perform(GraphRewrite event, EvaluationContext context, WindupConfigu
             final Map<String, CSVWriter> projectToFile = new HashMap<>();
             final Iterable<InlineHintModel> hints = hintService.findAll();
```
Comment (mrizzi) [inline reporting/impl/src/main/java/org/jboss/windup/reporting/export/ExportCSVFileRuleProvider.java]:
```suggestion
            final List<InlineHintModel> hints = hintService.findAll();
```

File: reporting/impl/src/main/java/org/jboss/windup/reporting/export/ExportCSVFileRuleProvider.java [hunk 7]
```diff
@@ -66,63 +70,108 @@ public void perform(GraphRewrite event, EvaluationContext context, WindupConfigu
...
             final Iterable<ClassificationModel> classifications = classificationService.findAll();
```
Comment (mrizzi) [inline reporting/impl/src/main/java/org/jboss/windup/reporting/export/ExportCSVFileRuleProvider.java]:
```suggestion
            final List<ClassificationModel> classifications = classificationService.findAll();
```

File: reporting/impl/src/main/java/org/jboss/windup/reporting/export/ExportCSVFileRuleProvider.java [hunk 8]
```diff
@@ -66,63 +70,108 @@ public void perform(GraphRewrite event, EvaluationContext context, WindupConfigu
...
+            List reportableEvents = new ArrayList<>();
```
Comment (mrizzi) [inline reporting/impl/src/main/java/org/jboss/windup/reporting/export/ExportCSVFileRuleProvider.java]:
```suggestion
            List<EffortReportModel> reportableEvents = new ArrayList<>();
```

File: reporting/impl/src/main/java/org/jboss/windup/reporting/export/ExportCSVFileRuleProvider.java [hunk 9]
```diff
@@ -66,63 +70,108 @@ public void perform(GraphRewrite event, EvaluationContext context, WindupConfigu
...
+            reportableEvents.addAll((List<InlineHintModel>)hints);
```
Comment (mrizzi) [inline reporting/impl/src/main/java/org/jboss/windup/reporting/export/ExportCSVFileRuleProvider.java]:
```suggestion
            reportableEvents.addAll(hints);
```

File: reporting/impl/src/main/java/org/jboss/windup/reporting/export/ExportCSVFileRuleProvider.java [hunk 10]
```diff
@@ -66,63 +70,108 @@ public void perform(GraphRewrite event, EvaluationContext context, WindupConfigu
...
+            reportableEvents.addAll((List<ClassificationModel>)classifications);
```
Comment (mrizzi) [inline reporting/impl/src/main/java/org/jboss/windup/reporting/export/ExportCSVFileRuleProvider.java]:
```suggestion
            reportableEvents.addAll(classifications);
```

File: reporting/api/src/main/java/org/jboss/windup/reporting/model/ApplicationReportModel.java [hunk 11]
```diff
@@ -154,4 +155,10 @@ default ProjectModel getProjectModel()
      */
     @MapInAdjacentProperties(label = "reportProperties")
     void setReportProperties(Map<String, String> map);
+
+    @Property(IS_EXPORTING_CSV)
+    Boolean getIsExportingCSV();
```
[the remark this answers is not in the document]
Comment (m-brophy) [reply reporting/api/src/main/java/org/jboss/windup/reporting/model/ApplicationReportModel.java]: Had to change this to getIsExportAllIssuesCSV() as the freemarker renderer got confused with whether isExportAllIssuesCSV was a method or a field and All Issues page failed to build

[the remark this answers is not in the document]
Comment (mrizzi) [reply reporting/api/src/main/java/org/jboss/windup/reporting/model/ApplicationReportModel.java]: Ah ok, i see :+1: 
Thanks for pointing this out, i didn't think about it.

File: reporting/impl/src/main/java/org/jboss/windup/reporting/export/ExportCSVFileRuleProvider.java (hunk 5 above)
Comment (mrizzi) [reply reporting/impl/src/main/java/org/jboss/windup/reporting/export/ExportCSVFileRuleProvider.java]: np :wink:

**Comment by mrizzi** [comment]
Thanks @m-brophy :+1: 
I'm going to build it locally again to test the latest changes.

File: reporting/impl/src/main/resources/reports/templates/migration-issues.ftl [hunk 12]
```diff
@@ -134,7 +134,11 @@
                 <div class="page-header page-header-no-border">
                     <h1>
                         <div class="main">${reportModel.reportName}
-                        <i class="glyphicon glyphicon-info-sign" data-toggle="tooltip" data-placement=right title="${reportModel.description}"></i></div>
+                        <i class="glyphicon glyphicon-info-sign" data-toggle="tooltip" data-placement=right title="${reportModel.description}"></i>
+                        <#if !(reportModel.projectModel??) && reportModel.isExportAllIssuesCSV>
+                            <a class="csvReport" style="font-size: 14pt;" align="right" href="../AllIssues.csv">(Download All Issues CSV)</a>
```
Comment (mrizzi) [inline reporting/impl/src/main/resources/reports/templates/migration-issues.ftl]:
```suggestion
                            <span style="float: right;">
                            <a class="csvReport" style="font-size: 14pt;" align="right" href="../AllIssues.csv">Download All Issues CSV</a>
                            </span>
```

Comment (m-brophy) [reply reporting/impl/src/main/resources/reports/templates/migration-issues.ftl]: Ah, sorry, I missed that, thanks. I'll check this in.

**Pull request merged**